test/functional/api/keymap_spec.lua

```diff
@@ -0,0 +1,243 @@
+  local function global_and_buffer_test(map,
+                                        option,
+                                        option_token,
+                                        global_on_result,
+                                        buffer_on_result,
+                                        global_off_result,
+                                        buffer_off_result,
+                                        new_windows)
+
+    local function make_new_windows(number_of_windows)
+      if new_windows == nil then
+        return nil
+      end
+
+      for _=1,number_of_windows do
+        nvim('command', 'new')
+      end
+    end
+
+    local mode = string.sub(map, 1,1)
+    -- Don't run this for the <buffer> mapping, since it doesn't make sense
+    if option_token ~= '<buffer>' then
+      it(string.format( 'returns %d for the key "%s" when %s is used globally with %s (%s)',
+          global_on_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' ' .. option_token .. ' foo bar')
+        local result = nvim('get_keymap', mode)[1][option]
+        eq(global_on_result, result)
+      end)
+    end
+
+    it(string.format('returns %d for the key "%s" when %s is used for buffers with %s (%s)',
+        buffer_on_result, option, option_token, map, mode), function()
+      make_new_windows(new_windows)
+      nvim('command', map .. ' <buffer> ' .. option_token .. ' foo bar')
+      local result = nvim('buf_get_keymap', 0, mode)[1][option]
+      eq(buffer_on_result, result)
+    end)
+
+    -- Don't run these for the <buffer> mapping, since it doesn't make sense
+    if option_token ~= '<buffer>' then
+      it(string.format('returns %d for the key "%s" when %s is not used globally with %s (%s)',
+          global_off_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' baz bat')
+        local result = nvim('get_keymap', mode)[1][option]
+        eq(global_off_result, result)
+      end)
+
+      it(string.format('returns %d for the key "%s" when %s is not used for buffers with %s (%s)',
+          buffer_off_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' <buffer> foo bar')
+
+        local result = nvim('buf_get_keymap', 0, mode)[1][option]
+        eq(buffer_off_result, result)
+      end)
+    end
+  end
+
+  -- Standard modes and backwards compatible
+  local mode_list = {'nnoremap', 'nmap', 'imap', 'inoremap', 'cnoremap'}
+  for mode in pairs(mode_list) do
+    global_and_buffer_test(mode_list[mode], 'silent', '<silent>', 1, 1, 0, 0)
+    global_and_buffer_test(mode_list[mode], 'nowait', '<nowait>', 1, 1, 0, 0)
+    global_and_buffer_test(mode_list[mode], 'expr', '<expr>', 1, 1, 0, 0)
+  end
+
+  -- noremap will now be 2 if script was used, which is not backwards compatible
```

Comment:
There was no API function previously, this is perfectly backwards compatible. Better state “which is different from maparg”.